### ISSUE TITLE
Allow admins to implement controller loops from the CLI (IFTTT)

### ISSUE THREAD

**Issue by smarterclayton** [kubernetes/kubernetes#5164]
It should be possible to script Kube from the cli via simple commands in kubectl that implement the list-and-watch pattern and pass those objects to scripts.  Examples

I want to watch all services and and tweak the external load balancer settings when it runs:

```
kubectl get services --watch --all -t "{{ .portalIP }}\n" | xargs ...
```

Needs something that makes it easy to exec a process per change, and potentially filter based on event type.

**Comment by smarterclayton** [comment]
In effect, it should be possible to create a real controller loop in bash only with a reasonably simple loop.  Maybe an --exec option that runs based on the item.

**Comment by bgrant0607** [comment]
Good idea. Related to #1899, as well as to scriptability discussed in cli-roadmap.md.

**Comment by smarterclayton** [comment]
In theory you then implement a controller in a pod by providing a bash script, and then you can guarantee it's always running on your cluster with an rc.

**Comment by you-n-g** [comment]
I'm a  prospective GSOC2015 participant. I've  finished some basic works such as learing go, reading the docs and so on.  I want to work on this issue to get started.
So I want to make sure that this issue is unassigned,  Is there anyone working on this issue?
Thanks!

**Comment by erictune** [comment]
Look at what `etcdctl` does.  (See https://github.com/coreos/etcd/tree/master/etcdctl)
It has a `--watch` options for humans and an `exec-watch` option for scripts, which calls a script with env vars set.

The one thing that I think could be improved is that the etcdctl can't do an atomic replacement of all state, so it can't handle re-lists atomically.

**Comment by erictune** [comment]
@bgrant0607  to reply on GSOC inquiry.

**Comment by smarterclayton** [comment]
No one is working on it at this time.

----- Original Message -----

**Comment by bgrant0607** [comment]
@you-n-g If you're looking for an easier getting-started project, try #2316.

**Comment by you-n-g** [comment]
@erictune  Really Grateful for your advices
@bgrant0607 Greatly thanks for spending time picking a  getting-started project for me.  I'll look into it #2316 first!

**Comment by bgrant0607** [comment]
#5511 has an example where a similar pattern would be useful.

**Comment by smarterclayton** [comment]
Started prototyping https://github.com/openshift/origin/pull/4196

**Comment by mml** [comment]
I like this idea as a starting point, but it will stop scaling up once the control loop needs to be made persistent, check for errors, retry, etc.  Once we have the CLI version implemented, do we have a model in mind for what we tell users when they get to that point?

**Comment by smarterclayton** [comment]
Good question.  80% of use cases I've walked over are fairly simple
(single loop, some resync).  Obviously this pattern works best for
maintaining simple invariants.  Kelsey and I talked a bit about a
"libkube" that bridged the gap for more complex function.  At the
limit, only a development team can really solve certain problems.  I
think that the way past the simple problem is the formalization of a
pattern, or a more powerful "controller library".  In combination with
the third party API, or the easily created server API framework we've
discussed, we do have a potential path out.

In practice, I suspect most organizations would solve most problems
with a simple check, get it integrated into their infra, and then wait
a year for it to be added to Kube.  A lot of the use cases we've
discussed so far are to integrate external systems into Kube, or do
simple policy based rules against namespaces or users.

**Comment by bgrant0607** [comment]
cc @thockin @lavalamp

**Comment by smarterclayton** [comment]
I have almost resumed working on this 5-6 times in the last two months.  Still haven't gotten over the hump

**Comment by lavalamp** [comment]
I think it shouldn't be too hard. I imagine a "watch-exec" which takes a
`--cmd` arg and passes as $1 one of sync, add, update, delete, and as $2
passes the object. Implement by starting an Informer. There's tons of other
ways to do it, too.

**Comment by smarterclayton** [comment]
I think the hard part of this is being familiar to a bash user with limited
kubernetes experience and giving them a simple stepwise process from

1. observe watch loop and get predictable behavior
2. handle more complex reaction (parallelism, retry)
3. handle delta resync

(i.e., up the complexity ladder).  Also handling tricky gotchas like bash
escaping, simple tooling that makes it easy to use API values, and how to
handle more complex scripting (if you want to use this with python script).

I'd want someone to be able to start using this and have it work (as long
as they understood which problem they are solving).  So the challenge isn't
so much in the mechanics as the experience, and educating users on why
controllers suck.

**Comment by erictune** [comment]
Echoing what Clayton just said: writing controllers is hard, and watch can give a false sense that it is easy.  We should encourage first-time controller writers to just poll.  Most people immediately get that when you poll, you might miss a transition and you have to handle seeing the same thing several times.  I think with watch they think it is more like a message queue, which it isn't.

**Comment by erictune** [comment]
Also, I recently built a controller in bash that used kubectl and polling.  It was not hard.  

I'd vote for closing this.

**Comment by smarterclayton** [comment]
Can you post the example so I can decide whether I agree with you for the
user I have in mind?

On Aug 17, 2016, at 4:19 AM, Eric Tune <notifications@github.com> wrote:

Also, I recently built a controller in bash that used kubectl and polling.
It was not hard.

I'd vote for closing this.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/5164#issuecomment-240345047>,
or mute the thread

**Comment by caruccio** [comment]
I'd use it to do my users billing. Today we are retrieving metric data from hawkular into a database. May not the the best, but it's the easier I could find at the time (any sugestions?)

Observer could inject nvironemtn variables into the script with basic information like KUBERNETES_NAMESPACE, KUBERNETES_OBJECT_NAME, etc...

**Comment by smarterclayton** [comment]
The first two arguments to the script are namespace and name, so you
have it.  You can record creation time, but for this loop you would
need to be aware that deletion could happen when you aren't running
(so you don't get exact deletion time).

You could also watch quota usage objects and sum them over time.

**Comment by caruccio** [comment]
What about implementing this loop on the master side, using some embeddable script language like lua or python?

Watching quota won't give detailed resource consumption. Does not works for me.

**Comment by smarterclayton** [comment]
We could just as easily add that to this command as a potential extension,
if we had reason.  We generally just don't put loops in the server side
because it should be just as easy to do it via the APIs.

Hrm - when you say detailed, what do you mean?  Watching quota would tell
you every byte granted to a user (more accurately than our other
mechanisms).  Agree you need additional data sources on what is actually
consumed.


On Sep 8, 2016, at 9:38 AM, Mateus Caruccio <notifications@github.com>
wrote:

What about implementing this loop on the master side, using some embeddable
script language like lua or python?

Watching quota won't give detailed resource consumption. Does not works for
me.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/5164#issuecomment-245599650>,
or mute the thread

**Comment by runseb** [comment]
Some feedback after trying this for 10 minutes.

So first I think it is great.

Now basic feedback:

* Downloading the openshift/origin image is heavy (500 MB) for getting a CLI.
* I faced issues writing a kubeconfig into that image as user 1001 does not exist. Quickly downloaded the binary release.
* For pure k8s, it feels weird downloading this from oc. I would prefer it to be a kubectl command (even as a proto).
* You need to open your firewall for incoming connections when `oc observe runs`. Feels strange...
* `oc observe service -- echo` only gets addition of services, first thing I did after creating a service and seeing the notification was to delete it ...and wait.
* `oc observe service --delete echo -- echo` receives addition and deletion , that's worth noting in the 5 minutes tryout README.
* I would like to run this server side in a deployment. So that definition needs to be available quickly and we even need a chart asap.
* Last thing I did was:

```
$ cat observe.py 
#!/usr/bin/env python

import sys
print sys.argv[1], sys.argv[2]
$ oc observe --all-namespaces services --delete ./observe.py -- ./observe.py
```

**Comment by caruccio** [comment]
@smarterclayton API client loops are great! I love API client loops! Lets make API client loops great again!  ...but they comes with a drawback: there is another component in the system as a whole, another point for failures and another stuff for Ops to maintain. We could reduce the risk with redundancy, which brings syncing in the game ("may I delete something twice?", "is this component already registered in my billing?")

Regarding quotas, I need to generate invoices with every single container start and stop timestamp, just like a phone bill. Hawkular gives me that. Does quotas do the same?

**Comment by smarterclayton** [comment]
Some feedback after trying this for 10 minutes.

So first I think it is great.

Now basic feedback:

   - Downloading the openshift/origin image is heavy (500 MB) for getting a
   CLI.


Yeah, it's a full Kube+OpenShift containerized distribution.  The client
binaries are smaller - we'll do a smaller image in the future.



   - I faced issues writing a kubeconfig into that image as user 1001 does
   not exist. Quickly downloaded the binary release.
   - For pure k8s, it feels weird downloading this from oc. I would prefer
   it to be a kubectl command (even as a proto).


We don't have a release structure for things outside of Kube core today -
rather than hacking things into a binary, I chose to reuse our existing
infra.



   - You need to open your firewall for incoming connections when oc
   observe runs. Feels strange...


In a pod?  Or just Docker?  Pods generally expect to be able to connect to
the master, so the expectation here is that your client can connect to the
master.  Should only be one way from client to API though.


   - oc observe service -- echo only gets addition of services, first thing
   I did after creating a service and seeing the notification was to delete it
   ...and wait.


By design - deletes are hard and are step 2.  To do delete properly you
need to guarantee you observe the deletes.  Rather than letting you cut
your fingers on the glass, it's opt in.  We can definitely explain that
better when the command starts.


   - oc observe service --delete echo -- echo receives addition and
   deletion , that's worth noting in the 5 minutes tryout README.


It should be in the CLI help for observe.


   - I would like to run this server side in a deployment. So that
   definition needs to be available quickly and we even need a chart asap.


Just use that image with kubectl run and add a secret (or give your service
account access to talk to the master via ABAC).  No chart needed.


   - Last thing I did was:

$ cat observe.py
#!/usr/bin/env python

import sys
print sys.argv[1], sys.argv[2]
$ oc observe --all-namespaces services --delete ./observe.py -- ./observe.py


There's maybe a longer term story for allowing you to get piped the object
instead of having to launch a child each time.  That is sufficiently more
complicated so we didn't want to cause pain.

Thanks for the feedback - please let us know if you try to build an
integration that is harder / more complex than you expect.

**Comment by smarterclayton** [comment]
On Sep 8, 2016, at 7:19 PM, Mateus Caruccio <notifications@github.com>
wrote:

@smarterclayton <https://github.com/smarterclayton> API client loops are
great! I love API client loops! Lets make API client loops great again!
...but they comes with a drawback: there is another component in the system
as a whole, another point for failures and another stuff for Ops to
maintain. We could reduce the risk with redundancy, which brings syncing in
the game ("may I delete something twice?", "is this component already
registered in my billing?")


Until we move to etcd3 this problem applies in many domains, not just
here.  Without a longer change history the window for "missed event" goes
up.  I definitely agree that that's something I would expect to work, and
being able to reliably monitor.

OTOH Kube is designed to keep things running.  If you run in a deployment
it can potentially benefit from monitoring the same as anything else.
Observe comes preloaded with a health and readiness check and a metrics
check.


Regarding quotas, I need to generate invoices with every single container
start and stop timestamp, just like a phone bill. Hawkular gives me that.
Does quotas do the same?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/5164#issuecomment-245771988>,
or mute the thread

**Comment by andrewklau** [comment]
This would be useful to replace the excruciatingly long nodejs scripts we've been writing. We are using WATCH to stream from the API and applying changes based on the data we receive. Our current scripts rely on redis to store temporary data and apply some logic to catch failures. We must also occasionally poll to ensure we don't miss any events.

Going against what was mentioned above, it would be nice to have `oc observe` provide an abstract queue method. This would mean we can do away with many error checking layers and simplify  integrations. It would be nice to turn a 600 line integration script into 5 lines!!

Kubernetes lets us do things in real time, integrating with kubernetes should also be in real time too. One other thing we also noticed was multiple events from the same action (perhaps OpenShift specific?) eg. creating a new project would send out multiple MODIFIED events to our scripts. Perhaps `oc observe` could hide these duplicate events.

@caruccio you can watch the changes field(s) eg. `data['object']['status']['total']['used']['limits.memory']` in resourcequotas, along with their timestamps. This needs to be done in real time though, so you can't go back to records like in Hawkular.

**Comment by smarterclayton** [comment]
> Going against what was mentioned above, it would be nice to have oc
> observe provide an abstract queue method. This would mean we can do away
> with many error checking layers and simplify integrations. It would be nice
> to turn a 600 line integration script into 5 lines!!
>

Just so I understand, when you say queue method, can you describe a bit
more what you mean?  You can view the command passed to observe as your
dequeue method - you won't miss events if you take too long to handle them,
for example.



> Kubernetes lets us do things in real time, integrating with kubernetes
> should also be in real time too. One other thing we also noticed was
> multiple events from the same action (perhaps OpenShift specific?) eg.
> creating a new project would send out multiple MODIFIED events to our
> scripts. Perhaps oc observe could hide these duplicate events.
>

Generally those aren't duplicates - they're individual modifications to the
object.  However, we can absolutely debounce those once they're converted
into arguments if you'd like (if previous execution resulted in the same
args, don't notify again).

**Comment by andrewklau** [comment]
> Just so I understand, when you say queue method, can you describe a bit
more what you mean?  You can view the command passed to observe as your
dequeue method - you won't miss events if you take too long to handle them,
for example.
> 

Sorry I was getting this confused with some queueing issues we had with doing this with nodejs and the streaming method. I'm not seeing this while trying `oc observe` which is great!

Can we start writing scripts, or would we be expecting major changes as this matures?

**Comment by bgrant0607** [comment]
This was the kubernetes-dev thread: https://groups.google.com/forum/#!topic/kubernetes-dev/Ot66oC_Il7E

**Comment by smarterclayton** [comment]
I would consider this an alpha that we are eager to get feedback on.  We
promise not to break openshift/observe:latest for at least 2 minor kube
releases, and unless this gets completely scrapped it's likely we can
create an equivalent.  We really need people finding value to commit to
keeping it.

**Comment by sebgoa** [comment]
Can we observe thirdpartyresources resource ?

Example:

I created a third-party resource called pinguin.

```
$ ./oc observe --all-namespaces thirdpartyresources -- echo
# 2016-10-05T10:03:25+02:00 Sync started
# 2016-10-05T10:03:25+02:00 Sync 93840	echo pin-guin.k8s.linuxcon.com
pin-guin.k8s.linuxcon.com
```

that's good, but now I would like to observe my _pinguins_:

```
 $ ./oc observe --all-namespaces pinguins -- echo
the server doesn't have a resource type "pinguins"
```

even though:

```
$ kubectl get pinguins
NAME      LABELS           DATA
crazy     linuxcon=rocks   {"apiVersion":"k8s.linuxcon.com/v1","kind":"PinGui...
```

**Comment by smarterclayton** [comment]
I believe you'd have to say pinguins.k8s.linuxcon.com

If that doesn't work let me know and I'll push a fix.

**Comment by sebgoa** [comment]
nope.

here is more info:

```
$ ./kubectl get thirdpartyresources
NAME                        DESCRIPTION                   VERSION(S)
pin-guin.k8s.linuxcon.com   A crazy pinguin at Linuxcon   v1
sebair: examples $ ./kubectl get pinguins
NAME      LABELS           DATA
crazy     linuxcon=rocks   {"apiVersion":"k8s.linuxcon.com/v1","kind":"PinGui...
sebair: examples $ ./kubectl get pinguins crazy -o yaml
apiVersion: k8s.linuxcon.com/v1
kind: PinGuin
metadata:
  creationTimestamp: 2016-10-05T07:38:34Z
  labels:
    linuxcon: rocks
  name: crazy
  namespace: default
  resourceVersion: "94052"
  selfLink: /apis/k8s.linuxcon.com/v1/namespaces/default/pinguins/crazy
  uid: b8a8cfe3-8ace-11e6-9891-069472000248
```

and the oc tries:

```
$ ./oc observe --all-namespaces thirdpartyresources -- echo
# 2016-10-05T18:08:47+02:00 Sync started
# 2016-10-05T18:08:47+02:00 Sync 93840	echo pin-guin.k8s.linuxcon.com
pin-guin.k8s.linuxcon.com
^C
sebair: Desktop $ ./oc observe --all-namespaces pinguins -- echo
the server doesn't have a resource type "pinguins"
sebair: Desktop $ ./oc observe --all-namespaces PinGuin -- echo
the server doesn't have a resource type "PinGuin"
sebair: Desktop $ ./oc observe --all-namespaces pinguins.k8s.linuxcon.com -- echo
the server doesn't have a resource type "pinguins" in group "k8s.linuxcon.com"
sebair: Desktop $ ./oc observe --all-namespaces pinguin.k8s.linuxcon.com -- echo
the server doesn't have a resource type "pinguin" in group "k8s.linuxcon.com"
sebair: Desktop $ ./oc observe --all-namespaces PinGuin.k8s.linuxcon.com -- echo
the server doesn't have a resource type "PinGuin" in group "k8s.linuxcon.com"
sebair: Desktop $ ./oc observe --all-namespaces pin-guin.k8s.linuxcon.com -- echo
the server doesn't have a resource type "pin-guin" in group "k8s.linuxcon.com"
```

**Comment by smarterclayton** [comment]
Thanks, will try to push a fix.

**Comment by tremble** [comment]
We've recently discovered "oc observe" and we love it.  The latest thing we're using it for is triggering a prune of build/deployment/image resources on a namespace when someone starts a build.

We've also used them for:
* Adding GlusterFS Service/Endpoints "just-in-time" to projects so that PVs can be properly mounted (this has allowed us to delete ~1600 Services) rather than having to add the objects to every Project.  This also allows us to automatically replace them when someone deletes them without knowing this will break their PVs
* Managing/Updating the various Quota objects based on updates to group membership
* Setting a default value for CronJob "concurrencyPolicy" (The default of 'Allow' has caused problems due to the lack of quotas on Jobs.  As such we only want to allow 'Replace' or 'Forbid' so we simply force 'Forbid' if it's not set to 'Replace' or 'Forbid')

Can't really offer much feedback other than a big thumbs up.
